src/Firestore/Query.php

```diff
@@ -0,0 +1,477 @@
+     */
+    private $query;
+
+    /**
+     * @var string?
+     */
+    private $transaction;
+
+    /**
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param array $query The Query object
+     * @param string|null The transaction ID, if the query is run in a transaction.
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        $parent,
+        array $query,
+        $transaction = null
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->parent = $parent;
+        $this->query = $query;
+        $this->transaction = $transaction;
+
+        if (!isset($this->query['from'])) {
+            throw new \InvalidArgumentException(
+                'Cannot instantiate a query which does not specify a collection selector (`from`).'
+            );
+        }
+    }
+
+    /**
+     * Get all documents matching the provided query filters.
+     *
+     * This method will not execute an immediate service call. The Query will be
+     * executed once you begin iterating over the rows (either by iterating
+     * directly on the QuerySnapshot, or by iterating over
+     * {@see Google\Cloud\Firestore\QuerySnapshot::rows()}).
+     *
+     * Example:
+     * ```
+     * $result = $query->snapshot();
+     * ```
+     *
+     * @param array $options Configuration options.
+     * @return QuerySnapshot
+     */
+    public function snapshot(array $options = [])
+    {
+        $options['maxRetries'] = (isset($options['maxRetries']))
+            ? $options['maxReties']
+            : FirestoreClient::MAX_RETRIES;
+
+        $call = function () use ($options) {
+            unset($options['maxRetries']);
+
+            return $this->connection->runQuery([
+                'parent' => $this->parent,
+                'structuredQuery' => $this->query,
+                'transaction' => $this->transaction,
+                'retries' => 0,
+            ] + $options);
+        };
+
+        return new QuerySnapshot($this->connection, $this->valueMapper, $this, $call, $options['maxRetries']);
+    }
+
+    /**
+     * Add a SELECT to the Query.
+     *
+     * Example:
+     * ```
+     * $query = $query->select(['firstName']);
+     * ```
+     *
+     * @param array $fieldPaths The projection to return, in the form of an
+     *        array of field paths. To only return the name of the document, use
+     *        `['__name__']`.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function select(array $fieldPaths)
+    {
+        $fields = [];
+        foreach ($fieldPaths as $field) {
+            $fields[] = [
+                'fieldPath' => $field
+            ];
+        }
+
+        return $this->newQuery([
+            'select' => [
+                'fields' => $fields
```

Comment:
You need to escape special characters in a field path (here and everywhere).

Reply:
This is implemented now using [`ValueMapper::escapeFieldPath()`](https://github.com/jdpedrie/google-cloud-php/blob/firestore/src/Firestore/ValueMapper.php#L112). Is it correct to only escape when given a `FieldPath`, or should I be attempting to escape dot-delimited strings as well?

Reply:
You need to escape dot-separated field paths as well. ``.`` should end up as ``\``.`\```.